src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLock.cs

```diff
@@ -0,0 +1,217 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Threading
+{
+    /// <summary>
+    /// A lightweight non-recursive mutex. Waits on this lock are uninterruptible (from Thread.Interrupt(), which is supported
+    /// in some runtimes). That is the main reason this lock type would be used over interruptible locks, such as in a
+    /// low-level-infrastructure component that was historically not susceptible to a pending interrupt, and for compatibility
+    /// reasons, to ensure that it still would not be susceptible after porting that component to managed code.
+    /// </summary>
+    internal sealed class LowLevelLock : IDisposable
+    {
+        // TODO: Tune these values
+        private const int SpinCount = 8;
+        private const int SpinSleep0Threshold = 4;
+
+        private const int LockedMask = 1;
+        private const int WaiterCountIncrement = 2;
+
+        // Layout:
+        //   - Bit 0: 1 if the lock is locked, 0 otherwise
+        //   - Remaining bits: Number of threads waiting to acquire a lock
+        private int _state;
+
+#if DEBUG
+        private Thread? _ownerThread;
+#endif
+
+        // Indicates whether a thread has been signaled, but has not yet been released from the wait. See SignalWaiter. Reads
+        // and writes must occur while _monitor is locked.
+        private bool _isAnyWaitingThreadSignaled;
+
+        private LowLevelSpinWaiter _spinWaiter;
+        private readonly Func<bool> _spinWaitTryAcquireCallback;
+        private readonly LowLevelMonitor _monitor;
+
+        public LowLevelLock()
+        {
+            _spinWaiter = default(LowLevelSpinWaiter);
+            _spinWaitTryAcquireCallback = SpinWaitTryAcquireCallback;
+            _monitor = new LowLevelMonitor();
+        }
+
+        ~LowLevelLock() => Dispose(false);
+        public void Dispose() => Dispose(true);
+
+        private void Dispose(bool disposing)
+        {
+            VerifyIsNotLockedByAnyThread();
+
+            if (disposing && _monitor != null)
+            {
+                _monitor.Dispose();
+            }
+
+            GC.SuppressFinalize(this);
+        }
+
+        public void VerifyIsLocked()
+        {
+#if DEBUG
```

Comment:
should the func defs all be under `#if DEBUG` since they have debug only code? Dont believe the compiler would optimize them out on non-debug?

Reply:
I wanted to attribute them with `[Conditional("DEBUG")]` but those methods are still compiled in release, so using debug-only variables like `_ownerThread` was yielding a compile error. I think they can still be attributed at such, just need an `#if DEBUG` around debug-only variable uses. Just wanted to avoid the nuisance of `#if'ing` out all calls to them. Even without the attribute, these would be empty methods in release build, the JIT should be able to optimize out those calls. In any case the attribute should get the best of both, as it looks like those methods don't even get IL generated for them in release builds.

Reply:
If you attribute them with `[Conditional("DEBUG")]`, there won't be any calls to them and the IL linker will strip the methods in the release binary.